### PR TITLE
[12.0] [MIG] account_invoice_supplier_self_invoice

### PR DESCRIPTION
**Pull request by Nikul-Chaudhary**
Account invoice purchase self invoice
=====================================

This module allows the configuration of the suppliers in order to define self invoices.
Self invoices are created by the customer in the name of the provider.
The provider must allow the customer to create them.

### CODE REVIEW

File: account_invoice_supplier_self_invoice/README.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+Images
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+
+* Enric Tobella <etobella@creublanca.es>
+* Nikul Chaudhary <nikulchaudhary2112@gmail.com>
```
Comment (etobella) [inline account_invoice_supplier_self_invoice/README.rst:51]: This kind of migration is not enough to add yourself as contributor.

**Review by etobella — CHANGES_REQUESTED** [review]
It is a standard migration.
Code Review OK, but I made a comment.
I just created a PR in order to improve the report #498 Maybe you can add it.

File: account_invoice_supplier_self_invoice/README.rst (hunk 1 above)
Comment (Nikul-Chaudhary) [reply account_invoice_supplier_self_invoice/README.rst:51]: @etobella It's Migrated module and tested all file for that, If I changed any singal file keep added your contributor.

Comment (pedrobaeza) [reply account_invoice_supplier_self_invoice/README.rst:51]: I think the contribution is enough for being here, but not on authors.

**Comment by Nikul-Chaudhary** [comment]
>  I just created a PR in order to improve the report #498 Maybe you can add it.

First this module is merged then after yours, so no need to added here.

File: account_invoice_supplier_self_invoice/README.rst (hunk 1 above)
Comment (Nikul-Chaudhary) [reply account_invoice_supplier_self_invoice/README.rst:51]: Yes @pedrobaeza , I added in only contribution section.

**Comment by etobella** [comment]
I know that there is no need. But it is an issue.

**Comment by Nikul-Chaudhary** [comment]
@etobella @OCA/accounting-maintainers could you please review code?

**Review by etobella — APPROVED** [review]

File: account_invoice_supplier_self_invoice/README.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+   Purchases'
+#. Go to 'Accounting/Invoicing > Purchases > Documents > Vendors Bills'
+#. Create an invoice for the provider and validate it
+#. The self invoice is accessible through the button: Self invoice
+#. You can create an invoice for the provider without the self invoice if you
+   uncheck 'Set self invoice'
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/159/11.0
```
Comment (LoisRForgeFlow) [inline account_invoice_supplier_self_invoice/README.rst]: Can you take the occasion to update to the new readme format?

**Comment by Nikul-Chaudhary** [comment]
@lreficent Done

**Comment by Nikul-Chaudhary** [comment]
@etobella @pedrobaeza @lreficent Could you please  review it?

File: account_invoice_supplier_self_invoice/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
```
Comment (Saran440) [inline account_invoice_supplier_self_invoice/models/res_partner.py]: For Python3, no need for utf-8 coding line as this is implicit.

File: account_invoice_supplier_self_invoice/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# © 2017 Creu Blanca
+# License AGPL-3.0 or later (https://www.gnuorg/licenses/agpl.html).
+
+{
+    "name": "Purchase Self Invoice",
+    "version": "12.0.2.0.0",
```
Comment (Saran440) [inline account_invoice_supplier_self_invoice/__manifest__.py]: version should be `12.0.1.0.0` for the first release of an 12.0 module.

File: account_invoice_supplier_self_invoice/views/report_self_invoice.xml [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    </template>
+    <report
+            id="account_invoices_self"
+            model="account.invoice"
+            string="Self Invoice"
+            report_type="qweb-pdf"
+            name="account_invoice_supplier_self_invoice.report_invoice_self"
+            file="account_invoice_supplier_self_invoice.report_invoice_self"
+            attachment_use="False"
+    />e
```
Comment (Saran440) [inline account_invoice_supplier_self_invoice/views/report_self_invoice.xml]:
```suggestion
    />
```

File: account_invoice_supplier_self_invoice/views/report_self_invoice.xml [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+</odoo>
```
Comment (Saran440) [inline account_invoice_supplier_self_invoice/views/report_self_invoice.xml]:
```suggestion
</odoo>

```

File: account_invoice_supplier_self_invoice/views/report_self_invoice.xml [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                        <tr class="border-black">
+                            <td>
+                                <strong>Subtotal</strong>
+                            </td>
+                            <td class="text-right">
+                                <span t-field="o.amount_untaxed"
+                                      t-options='{"widget": "monetary", "display_currency": o.currency_id}'/>
+                            </td>
+                        </tr>
+                        <t t-foreach="o._get_tax_amount_by_group()"
```
Comment (Saran440) [inline account_invoice_supplier_self_invoice/views/report_self_invoice.xml]: Error this line, Can't find function  `_get_tax_amount_by_group()` in `account.invoice`

**Comment by Nikul-Chaudhary** [comment]
@Saran440 Code is Done As per your suggestion.

File: account_invoice_supplier_self_invoice/readme/USAGE.rst [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#. Go to 'Accounting/Invoicing > Purchases > Master Data > Vendors'
```
Comment (Saran440) [inline account_invoice_supplier_self_invoice/readme/USAGE.rst]: Please update menu in V12

**Review by Saran440 — COMMENTED** [review]
@Nikul-Chaudhary when I click button Self Invoice -> Error :  `IndexError: list index out of range`

**Comment by Nikul-Chaudhary** [comment]
@Saran440 I Updated code as per your comment.

And Check in Runbot Work Good, Could you please check again.

**Review by jarroyomorales — APPROVED** [review]

**Comment by Saran440** [comment]
@Nikul-Chaudhary Sorry, I mean button Self Invoice on Vendor bills
Selection_006

Selection_007

and my config set self invoice. Did I do something wrong?
Selection_008

File: account_invoice_supplier_self_invoice/models/account_invoice.py [hunk 9]
```diff
@@ -48,7 +48,14 @@ def create(self, vals):
                         )
         return super(AccountInvoice, self).create(vals)
 
-    @api.multi
-    def action_view_account_invoice_self(self):
+    @api.model
+    def action_view_account_invoice_self(self, docids, data=None):
         self.ensure_one()
```
Comment (jarroyomorales) [inline account_invoice_supplier_self_invoice/models/account_invoice.py]: I think the ensure one is not necessary here

**Comment by jarroyomorales** [comment]
@OCA/accounting-maintainers Can this be merged?

**Comment by tafaRU** [comment]
@Nikul-Chaudhary could you please squash last commits to the migration commit?

**Comment by Nikul-Chaudhary** [comment]
@tafaRU @jarroyomorales Squashed commits

**Comment by jarroyomorales** [comment]
@OCA/accounting-maintainers Can this be merged?

File: account_invoice_supplier_self_invoice/models/account_invoice.py (hunk 9 above)
Comment (etobella) [reply account_invoice_supplier_self_invoice/models/account_invoice.py]: Why do you add the @api.model here?

File: account_invoice_supplier_self_invoice/models/account_invoice.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                getattr(invoice, onchange_method)()
+                for field in changed_fields:
+                    if field not in vals and invoice[field]:
+                        vals[field] = invoice._fields[field].convert_to_write(
+                            invoice[field], invoice
+                        )
+        return super(AccountInvoice, self).create(vals)
+
+    @api.model
+    def action_view_account_invoice_self(self, docids, data=None):
```
Comment (jarroyomorales) [inline account_invoice_supplier_self_invoice/models/account_invoice.py:52]: maybe it is better to set @api.multi and remove docids (docids will be self.ids)

File: account_invoice_supplier_self_invoice/models/res_partner.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    )
+
+    def set_self_invoice(self):
+        for record in self:
+            record.self_invoice = not record.self_invoice
+            if record.self_invoice_sequence_id:
+                continue
+            if record.self_invoice:
+                record.self_invoice_sequence_id = \
+                    self.env['ir.sequence'].create({
```
Comment (jarroyomorales) [inline account_invoice_supplier_self_invoice/models/res_partner.py:33]: Maybe a sudo is required, in order to create the sequence. Most users are not allowed to create sequences.

**Review by jarroyomorales — CHANGES_REQUESTED** [review]
Some comments. Can you review them?

File: account_invoice_supplier_self_invoice/models/account_invoice.py (hunk 10 above)
Comment (jarroyomorales) [reply account_invoice_supplier_self_invoice/models/account_invoice.py:52]: This function is not working, could you change it for this?

``` python
    @api.multi
    def action_view_account_invoice_self(self):
        report = self.env['ir.actions.report']._get_report_from_name(
            'account_invoice_supplier_self_invoice.report_invoice_self')
        return report.report_action(self)
```

**Comment by jarroyomorales** [comment]
@Nikul-Chaudhary Could you check comments? Thank you! :smile:

**Comment by AaronHForgeFlow** [comment]
Superseded here https://github.com/OCA/account-invoicing/pull/796 and merged :tada:  Closing this one.

**Pull request closed without merging**